src/sources/kubernetes_logs/parser/docker.rs

```diff
@@ -0,0 +1,133 @@
+use crate::{
+    event::{self, Event, LogEvent, Value},
+    transforms::{
+        json_parser::{JsonParser, JsonParserConfig},
+        Transform,
+    },
+};
+use chrono::{DateTime, Utc};
+use lazy_static::lazy_static;
+use snafu::{OptionExt, ResultExt, Snafu};
+use string_cache::DefaultAtom as Atom;
+
+pub fn build() -> Box<dyn Transform> {
+    Box::new(Docker::new())
+}
+
+lazy_static! {
+    pub static ref TIME: Atom = Atom::from("time");
+    pub static ref LOG: Atom = Atom::from("log");
+}
+
+/// Parser for the docker log format.
+///
+/// Expects logs to arrive in a JSONLines format with the fields names and
+/// contents specific to the implementation of the Docker `json` log driver.
+///
+/// Normalizes parsed data for consistency.
+#[derive(Debug)]
+pub struct Docker {
+    json_parser: JsonParser,
+}
+
+impl Docker {
+    /// Create a new [`Docker`] parser.
+    pub fn new() -> Self {
+        let json_parser = {
+            let mut config = JsonParserConfig::default();
+            config.drop_field = true;
+
+            // Drop so that it's possible to detect if message is in json format.
+            config.drop_invalid = true;
+
+            config.into()
+        };
+
+        Self { json_parser }
+    }
+}
+
+impl Transform for Docker {
+    fn transform(&mut self, event: Event) -> Option<Event> {
+        let mut event = self.json_parser.transform(event)?;
+        normalize_event(event.as_mut_log()).ok()?;
+        Some(event)
+    }
+}
+
+fn normalize_event(log: &mut LogEvent) -> Result<(), NormalizationError> {
+    // Parse and rename timestamp.
+    let time = log.remove(&TIME).context(TimeFieldMissing)?;
+    let time = match time {
+        Value::Bytes(val) => val,
+        _ => return Err(NormalizationError::TimeValueUnexpectedType),
+    };
+    let time = DateTime::parse_from_rfc3339(String::from_utf8_lossy(time.as_ref()).as_ref())
+        .context(TimeParsing)?;
+    log.insert(
+        event::log_schema().timestamp_key(),
+        time.with_timezone(&Utc),
+    );
+
+    // Parse message, remove trailing newline and detect if it's partial.
+    let message = log.remove(&LOG).context(LogFieldMissing)?;
+    let mut message = match message {
+        Value::Bytes(val) => val,
+        _ => return Err(NormalizationError::LogValueUnexpectedType),
+    };
+    let is_partial = if message.last().map(|&b| b as char == '\n').unwrap_or(false) {
```

Comment:
Turns out this is bugged!
In fairness, the interface for detecting partial events in docker is garbage - no formal interface exists, so we can only apply heuristics.
Partial messages are only split in docker at the maximum message length. Thus, for a message to be partial it also has to have a certain length. Now, whether that message will or won't actually be partial if it has exactly the max length is unknown. We probably should consider all messages with the exact length of `16 * 1024` bytes partial by-defailt, and then, if they end with newline - consider that an exception and make them non-partial. Logic is still broken, but bettern than what we currently have.

Thoughts?

I found this bug via e2e tests :smile:

Reply:
That sounds awful, and that workaround sounds sufficient. Do you think you could note this better in the code?

Reply:
Definitely. This is one of those non-trivial pieces of logic that needs proper documentation, rationale and tests.